### PR TITLE
T20853: Break the ekncontent dependency and use EknServices

### PR DESCRIPTION
**Pull request by smspillaz**
Previously we read content data from ekncontent directly. Amongst
other things, this was bad because it tied us to a single SDK version
because we can't load more than one ekncontent library at a time. This
commit breaks that dependency by using EknServices to perform queries
instead. It uses the ArticleMetadata query interface in the corresponding
pull rquest (endlessm/eos-knowledge-services#38)
to perform queries in a similar way that we performed queries against
eknservices directly, except that data is marshalled through D-Bus as
opposed to a binary interface.

Updated the tests to use a fake EknServicesContentDbConnection class
which implemnets a basic subset of ekncontent functionality against
a "shard" implemented in a python dict. This mean that we could drop
a lot of the shard generation code, which is nice.

https://phabricator.endlessm.com/T20853

### CODE REVIEW

**Comment by smspillaz** [comment]
For what its worth, I have a branch (https://github.com/endlessm/eos-companion-app-integration/tree/T20853-rebase-T21272-integrate) to rebase this one on top of T21272, since it looks like we will end up merging that one first.

**Comment by smspillaz** [comment]
A quick heads up that this has now been rebased on #77

**Comment by cosimoc** [comment]
@smspillaz I think it would be nice to get this another rebase now that #77 has been merged to master.

**Comment by cosimoc** [comment]
>@smspillaz I think it would be nice to get this another rebase now that #77 has been merged to master.

Just did that myself, will take another final look now.

File: eoscompanion/service.py [hunk 1]
```diff
@@ -29,15 +29,25 @@
...
 
-    def __init__(self, application, port, *args, **kwargs):
-        '''Initialize the service and create webserver on port.'''
+    def __init__(self, application, port, content_db_query, *args, **kwargs):
+        '''Initialize the service and create webserver on port.
+
+        :content_db_query: is a function satisfying the signature
+                           def content_db_query(app_id: str,
+                                                query: dict,
+                                                callback: GAsyncReadyCallback )
```
Comment (cosimoc) [inline eoscompanion/service.py]: Nit: extra space before closing paren here

File: test/test_service.py [hunk 2]
```diff
@@ -303,13 +302,20 @@ def local_endpoint(port, endpoint, version='v1'):
...
+VIDEO_APP_THUMBNAIL_EKN_ID = 'videofilethumbnail'
+CONTENT_APP_THUMBNAIL_EKN_ID = 'thumbnailforarticles'
+CONTENT_APP_EMBEDDED_IMAGE_EKN_ID = 'embeddedimage'
 
 VIDEO_APP_FAKE_CONTENT = (
     'Not actually an MPEG-4 file, just a placeholder for tests\n'
 )
-
+EMBEDDED_IMAGE_FAKE_CONTENT = (
+    'Not actually an embedded iamge, just a placeholder\n'
```
Comment (cosimoc) [inline test/test_service.py]: Typo: `iamge`

File: eoscompanion/routes.py [hunk 3]
```diff
@@ -1184,29 +1191,30 @@ def _on_all_searches_complete(search_results):
...
 
-                try:
-                    query_results = engine.query_finish(result)
-                    all_models.extend([
-                        ApplicationModel(app_id=applications[index].app_id,
-                                         model=m)
-                        for m in query_results.get_models()
-                    ])
-                except GLib.Error as error:
+                if error != None:
```
Comment (cosimoc) [inline eoscompanion/routes.py]: Should use `is not` instead of `!=` here, no?

File: eoscompanion/routes.py [hunk 4]
```diff
@@ -950,36 +953,35 @@ def _on_got_metadata_callback(load_metadata_error, metadata_bytes):
...
-			                          search_terms=search_term)
-    Eknc.Engine.get_default().query(query, None, callback)
+    content_db_conn.query(app_id,
+                          query={
+                              'tags-match-any': tags or [
+                                  'EknArticleObject',
+                                  'EknSetObject'
+                              ],
+                              'limit': limit or _SENSIBLE_QUERY_LIMIT,
+                              'offset': offset,
```
Comment (cosimoc) [inline eoscompanion/routes.py]: Should this not be `offset or 0` like in the previous code, or does `query()` take care of that?

File: eoscompanion/routes.py [hunk 5]
```diff
@@ -897,50 +871,79 @@ def app_id_to_runtime_version(app_id):
...
+                runtime_name_for_app_id_cached(query['applicationId'])
+            )
+            msg.set_status(Soup.Status.OK)
+            json_response(msg, {
+                'status': 'ok',
+                'payload': metadata_json
+            })
+            server.unpause_message(msg)
+
+        if shards_error != None:
```
Comment (cosimoc) [inline eoscompanion/routes.py]: `is not None`?

File: eoscompanion/routes.py [hunk 6]
```diff
@@ -651,225 +633,217 @@ def companion_app_server_content_data_route(server, msg, path, query, context):
...
+            # depending on whether it needs to be converted.
+            conditionally_wrap_blob_stream(blob,
+                                           content_type,
+                                           query,
+                                           content_metadata,
+                                           content_db_conn,
+                                           shards,
+                                           _on_got_wrapped_stream)
+
+        if shards_error != None:
```
Comment (cosimoc) [inline eoscompanion/routes.py]: `is not None`

File: eoscompanion/routes.py [hunk 7]
```diff
@@ -651,225 +633,217 @@ def companion_app_server_content_data_route(server, msg, path, query, context):
...
-        'applicationId={applicationId}, contentId={contentId}'.format(
-            contentId=query['contentId'],
-            applicationId=query['applicationId'],
-            clientId=query['deviceUUID']
-        )
-    )
+                Note that this is not a GAsyncReadyCallback, we instead get
+                a tuple of an error or a stream and length.
+                '''
+                if error != None:
```
Comment (cosimoc) [inline eoscompanion/routes.py]: `is not None`

File: eoscompanion/routes.py [hunk 8]
```diff
@@ -464,115 +466,90 @@ def _on_loaded_application_colors(_, result):
...
+    del context
 
-    def _callback(_, result):
+    def _callback(error, result):
         '''Callback function that gets called when we are done.'''
-        try:
-            query_results = engine.query_finish(result)
-            models = query_results.get_models()
-
+        if error != None:
```
Comment (cosimoc) [inline eoscompanion/routes.py]: `is not None`

File: eoscompanion/routes.py [hunk 9]
```diff
@@ -464,115 +466,90 @@ def _on_loaded_application_colors(_, result):
...
-                    'error': serialize_error_as_json_object(
-                        EosCompanionAppService.error_quark(),
-                        EosCompanionAppService.Error.FAILED,
-                        detail={
-                            'server_error': str(error)
-                        }
-                    )
-                })
-
+        if error != None:
```
Comment (cosimoc) [inline eoscompanion/routes.py]: `is not None`

File: eoscompanion/content_adjusters.py [hunk 10]
```diff
@@ -124,14 +123,13 @@ def render_mobile_wrapper(renderer,
     of outgoing links to internal links. Then we inject some javascript which
     at browser-render time, rewrites the page to resolve all those links.
     '''
-    def _on_queried_sets(src, result):
+    def _on_queried_sets(error, result):
         '''Called when we finish querying set objects.'''
-        try:
-            results = src.query_finish(result)
-        except GLib.Error as error:
+        if error != None:
```
Comment (cosimoc) [inline eoscompanion/content_adjusters.py]: `is not None`

File: eoscompanion/routes.py (hunk 4 above)
Comment (smspillaz) [reply eoscompanion/routes.py]: Good point. I am not sure if passing `None` to the d-bus interface will get silently ignored, so better to use `offset or 0`

File: eoscompanion/content_adjusters.py (hunk 10 above)
Comment (smspillaz) [reply eoscompanion/content_adjusters.py]: Ah, missed this one

File: eoscompanion/eknservices_bridge.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+    EosShard,
+    Gio,
+    GLib
+)
+
+from .functional import all_asynchronous_function_calls_closure
+
+
+_EKNSERVICES_DBUS_NAME = 'com.endlessm.EknServices2.SearchProviderV2'
+_EKNSERVICES_BASE_DBUS_PATH = '/com/endlessm/EknServices2/SearchProviderV2'
```
[the remark this answers is not in the document]
Comment (smspillaz) [reply eoscompanion/eknservices_bridge.py]: Yeah, we'll need code that figures out the right name and path to use once EknServices3 is available. For now, there is no change required.

**Pull request merged**